identity/keystore_filesystem.go

```diff
@@ -18,17 +18,189 @@
 package identity
 
 import (
-	"github.com/ethereum/go-ethereum/accounts/keystore"
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/rand"
+	"crypto/sha512"
+	"errors"
+	"io"
+	"sync"
+
+	"github.com/awnumar/memguard"
+	"github.com/ethereum/go-ethereum/accounts"
+	ethKs "github.com/ethereum/go-ethereum/accounts/keystore"
+	"github.com/ethereum/go-ethereum/common"
 	"github.com/rs/zerolog/log"
+	"golang.org/x/crypto/hkdf"
 )
 
-// NewKeystoreFilesystem create new keystore, which keeps keys in filesystem
-func NewKeystoreFilesystem(directory string, lightweight bool) *keystore.KeyStore {
+// NewKeystoreFilesystem create new keystore, which keeps keys in filesystem.
+func NewKeystoreFilesystem(directory string, lightweight bool) *Keystore {
+	var ks *ethKs.KeyStore
 	if lightweight {
 		log.Debug().Msg("Using lightweight keystore")
-		return keystore.NewKeyStore(directory, keystore.LightScryptN, keystore.LightScryptP)
+		ks = ethKs.NewKeyStore(directory, ethKs.LightScryptN, ethKs.LightScryptP)
+	} else {
+		log.Debug().Msg("using heavyweight keystore")
+		ks = ethKs.NewKeyStore(directory, ethKs.StandardScryptN, ethKs.StandardScryptP)
+	}
+
+	return &Keystore{
+		ethKeystore: ks,
+		derivedKeys: make(map[common.Address]*memguard.Enclave),
+	}
+}
+
+// Keystore handles everything that's related to eth accounts.
+type Keystore struct {
+	ethKeystore *ethKs.KeyStore
+
+	derivedKeys    map[common.Address]*memguard.Enclave
+	derivedKeyLock sync.Mutex
+}
+
+// Accounts returns all accounts.
+func (ks *Keystore) Accounts() []accounts.Account {
+	return ks.ethKeystore.Accounts()
+}
+
+// NewAccount creates a new account.
+func (ks *Keystore) NewAccount(passphrase string) (accounts.Account, error) {
+	return ks.ethKeystore.NewAccount(passphrase)
+}
+
+// Find finds an account.
+func (ks *Keystore) Find(a accounts.Account) (accounts.Account, error) {
+	return ks.ethKeystore.Find(a)
+}
+
+// Unlock unlocks an account.
+func (ks *Keystore) Unlock(a accounts.Account, passphrase string) error {
+	err := ks.ethKeystore.Unlock(a, passphrase)
+	if err != nil {
+		return err
+	}
+
+	derived, err := ks.deriveKey(a, passphrase)
```

Comment:
Is this expensive operation? Can we deriveKey just before encrypting / decrypting and destroy (wipe) right after we used it?

Reply:
It's expensive in a sense that we need to store the pk in memory if we want to invoke this. I'd rather save the derived key.

Reply:
It's expensive in a sense that we need to store the pk in memory if we want to invoke this. I'd rather save the derived key.

Reply:
Agree, if we can rely on memguard doing its job properly.